Readme.md

```diff
@@ -0,0 +1,243 @@
+
+The file to edit is located in `af/shovel/centrifugation.py`.
+
+2. Push a new docker image
+
+You should then build a new docker image by bumping the version number inside of
+`af/shovel/build`.
+
+3. Redeploy an update
+
+To redeploy the updated version of the pipeline you should run from the
+`ansible` directory of the [ooni-sysadmin
+repo](https://github.com/thetorproject/ooni-sysadmin) the following:
+
+```
+./play deploy-pipeline.yml --list-tags airflow-dags
+```
+
+4. Reprocess the data
+
+To re-process the data through the Airflow UI and scheduler, do the following:
+
+* Got to Tree view
+
+* Click on the task you want to reprocess
+
+* Click on "Clear" and select "Past"
+
+* Click "Done"
+
+* In the confirmation page click on "OK"
+
+Note: the airflow scheduler is a bit imperfect and has issues handling too many
+dependent tasks.
+
+In the future we will have a `Makefile` to avoid airflow scheduling overhead.
+
+## How do add fingerprints to database
+
+Fingerprints are what we use to detect web censorship and we currently only
+consider censorship to be "confirmed" when we have a positive ID on the
+blockpage fingerprint.
+
+In here we explain how you can go about adding new fingerprints to the DB.
+
+1. Create the SQL migration scripts
+
+You should use a template the migration file
+`af/oometa/003-fingerprints.install.sql` and
+`af/oometa/003-fingerprints.rollback.sql`.
+
+Based on this create a new set of files in `af/oometa/` called
+`00x-fingerprints.install.sql` and `00x-fingerprints.rollback.sql`.
+
+2. Run the migration script on the DB
+
+You should then manually apply the schema by running the `install` sql script on
+the database directly.
+
+3. Bump version information in `centrifugation.py`
+
+Look inside of `centrifugation.py` and bump `CODE_VER`.
+In the same file look for references to `compat_code_ver` and change that where
+needed (you will probably have to append to the list everywhere).
+
+4. Fix assertion to the fingerprint table
+
+At some point inside of `centrifugation.py` there is a line that says something like:
+
+```
+assert '916446978a4a86741d0236d19ce7157e' == next(c)[0], 'fingerprint table does not match CODE_VER
+```
+
+You should change the hash value to the output of the following query on the DB:
+
+```
+select md5(
+    md5(array_agg(fingerprint_no    order by fingerprint_no)::text) ||
+    md5(array_agg(origin_cc         order by fingerprint_no)::text) ||
+    md5(array_agg(body_substr       order by fingerprint_no)::text) ||
+    md5(array_agg(header            order by fingerprint_no)::text) ||
+    md5(array_agg(header_prefix     order by fingerprint_no)::text) ||
+    md5(array_agg(header_value      order by fingerprint_no)::text) ||
+    ''
+) from fingerprint
+```
+
+5. Deploy and re-process the data
+
+Basically you need to do the same thing you did in "Modifying the centrifugation
+stage" from step 2. onwards.
+
+The task you want to re-run is labelled `meta_pg`.
```

Comment:
```diff
+Reprocessing of whole dataset takes couple of days, it's done asynchronously,
+but the pipeline does not implement any priorities, so it may block data ingestion for a while.
```